src/data/fin/succ_pred.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2022 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import order.succ_pred.basic
+
+/-!
+# Successors and predecessors of `fin n`
+
+In this file, we show that `fin n` is both a `succ_order` and a `pred_order`. Note that they are
+also archimidean, but this is derived from the general instance for well-orderings as opposed
+to a specific `fin` instance.
+
+-/
+
+namespace fin
+
+instance : ∀ {n : ℕ}, succ_order (fin n)
+| 0 := by constructor; exact elim0
+| (n+1) :=
+{ succ := λ i, if i < fin.last n then i + 1 else i,
+  le_succ := λ a,
+  begin
+    cases n,
+    { exact (subsingleton.elim _ _).le },
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      exact (a : ℕ).le_succ },
+    { exact le_rfl }
+   end,
+  max_of_succ_le := λ a,
+  begin
+    split_ifs,
+    { rw [le_iff_coe_le_coe, coe_add_one_of_lt h],
+      intro h,
+      cases (lt_add_one _).ne (antisymm (a : ℕ).le_succ h) },
+    { rintro -,
+      rw [eq_last_of_not_lt h],
+      exact is_max_top }
+  end,
+  succ_le_of_lt := λ a b h', let h := h'.trans_le (le_last _) in
+                             by rwa [if_pos h, le_iff_coe_le_coe, coe_add_one_of_lt h],
+  le_of_lt_succ := λ a b h,
+  begin
+    split_ifs at h with hab,
+    swap, { exact h.le },
+    rw [lt_iff_coe_lt_coe, coe_add_one_of_lt hab] at h,
+    exact nat.le_of_lt_succ h
+  end }
+
+@[simp] lemma succ_eq {n : ℕ} : succ_order.succ = λ a, if a < fin.last n then a + 1 else a := rfl
+lemma succ_apply {n : ℕ} (a) : succ_order.succ a = if a < fin.last n then a + 1 else a := rfl
```

Comment:
Because this is an applied lemma, it should get priority hence be in simp normal form.
```suggestion
@[simp] lemma succ_apply {n : ℕ} (a) : succ_order.succ a = if a < fin.last n then a + 1 else a := rfl
```

Reply:
hmm, why would we want this? I can see either the argument that `simp` should not remove the `succ` at all (that way we can apply `succ_order` lemmas) or we want it removed all the time (which is why I thought that the non-applied versions were better for simp).

Reply:
Something about priority... If you tag the non applied lemma with `simp`, you might as well tag the applied one. Sorry, I don't know more of the details!